keps/sig-network/4969-cluster-domain-downward-api/README.md

```diff
@@ -0,0 +1,930 @@
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+The Pod `foo` needs to access its sibling Service `bar` in the same namespace.
+It adds two `env` bindings:
+
+``` yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: foo
+spec:
+  containers:
+    - name: foo
+      env:
+        - name: NAMESPACE
+          valueFrom:
+            fieldRef: metadata.namespace
+        - name: CLUSTER_DOMAIN
+          valueFrom:
+            clusterPropertyRef: clusterDomain
+```
+
+`foo` can now perform the query by running `curl http://bar.$NAMESPACE.svc.$CLUSTER_DOMAIN/`.
+
+(Of course, in practice this would likely be integrated into the app itself, not by shelling into bash, but the principle still applies.)
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Split Domains
+
+Kubernetes currently does not prohibit different kubelets from specifying 
+different cluster domains (`node-a` could set `cluster.local` while `node-b` 
+specifies `cluster.remote`).
+Exposing FQDNs generated using this API could cause issues in these mixed
+environments, since `node-b` might not be able to resolve `cluster.local`
+FQDNs correctly.
+
+For this KEP to make sense, this would have to be explicitly prohibited.
+
+## Design Details
+
+A new Downward API `clusterPropertyRef: clusterDomain` would be introduced, which can be projected into an environment variable or a volume file.
+
+<<[UNRESOLVED @nightkr @aojea @thockin]>>
+The name is undecided. Other candidates:
+
+- `nodePropertyRef` (@aojea)
+- `runtimeConfigs` (@thockin)
+
+This also implies a decision about who "owns" the setting, the cluster as a whole or the individual kubelet.
```

Comment:
That actually seems orthogonal to exposing this to the pods. We could expose the service domain as reported to kubelet without requiring this.

Reply:
Yeah.. this was brought up before in https://github.com/kubernetes/enhancements/pull/4972#discussion_r1853883953. At this point I think it makes sense to drop the paragraph.